src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -461,7 +491,10 @@ private void EnsureCapacitySlow(int capacity)
             if (_propertyStorage != null)
             {
                 var storage = _propertyStorage;
-                capacity = Math.Max(storage.Properties.Length, capacity);
+                
+                // If we're converting from properties, it's likely due to an 'add' to make sure we have at least
+                // the default amount of space.
+                capacity = Math.Max(DefaultCapacity, Math.Max(storage.Properties.Length, capacity));
```

Comment:
I found this issue while testing the new `TryAdd` method. We want to always have a minimum of 4. The properties -> array path could cause you to end up with an array of size 1, which would get resized to 2 next time you add... you get the idea.